File: bridge/submissions/accepted/binarysearch.py
def solve(B: int, N: int, S: list) -> int:
    low = min(S)
    high = max(S)
    while low < high:
        mid = (high + low) // 2
        cost = 0
        danger = 0
        for h in S:
            danger += max(mid - h, 0)
            cost += max(h - mid, 0)
        if cost > B:
            low = mid + 1
        else:
            high = mid
    return low


def main():
    T = int(input())
    for _ in range(T):
        temp = input().split()
        B, N = int(temp[0]), int(temp[1])
        S = [int(x) for x in input().split()]
        print(solve(B, N, S))


if __name__ == "__main__":
    main()

File: bridge/submissions/accepted/binarysearch.cpp
#include <bits/stdc++.h>
using namespace std;

/**
 * Return the height H in which the danger is minimized and satisfies the budget constraints.
 *
 * B: a non-negative integer
 * N: a positive integer
 * S: an array of N integers
 */
long long solve(long long B, int N, long long S[]) {
    long long low = LLONG_MAX;
    long long high = 0;
    for (int i = 0; i < N; ++i) {
        low = min(low, S[i]);
        high = max(high, S[i]);
    }
    while (low < high) {
        long long mid = (low + high) / 2;
        long long cost = 0;
        long long danger = 0;
        for (int i = 0; i < N; i++) {
            if (S[i] < mid) {
                danger += mid - S[i];
            } else {
                cost += S[i] - mid;
            }
        }

        if (cost > B) {
            low = mid + 1;
        } else {
            high = mid;
        }
    }
    return high;
}

int main() {
    int T;
    cin >> T;
    for (int i = 0; i < T; i++) {
        long long B;
        int N;
        cin >> B >> N;
        long long S[N];
        for (int j = 0; j < N; j++) {
            cin >> S[j];
        }
        cout << solve(B, N, S) << '\n';
    }
}
